### PR TITLE
[Windows] Improve SCM error logging

### PR DESCRIPTION
**Pull request by davinci26**
Signed-off-by:  Sotiris Nanopoulos <sonanopo@microsoft.com>

Commit Message:

Envoy registers with SCM before it parses the command line and as a result we don't have access to loggers. Additionally SCM by default captures the stdout/stderr so we can't really write to the standard streams. As a result debugging startup failures in SCM is a bit tricky.

To solve this I used the `spdlog::sinks::win_eventlog_sink_mt` which writes to the Windows Event Viewer. We use this logger directly for failures during SCM initialization.

Additional Description: N/A
Risk Level: Low
Testing: Manual
Docs Changes: Done
Release Notes: N/A
Platform Specific Features: Windows Only

### CODE REVIEW

**Comment by davinci26** [comment]
@wrowe  PTAL

**Comment by davinci26** [comment]
@wrowe lets get it merged before 1.18 since it is an already experimental feature. I look for your comments

**Review by wrowe — DISMISSED** [review]
This looks good to me, ready for a pass by @envoyproxy/senior-maintainers specific to logging architecture. Since this gets the error text of well-formed exceptions into the event log for troubleshooting, it's a good step forward.

File: source/exe/win32/service_base.cc [hunk 1]
```diff
@@ -50,8 +50,11 @@ bool ServiceBase::TryRunAsService(ServiceBase& service) {
     if (last_error == ERROR_FAILED_SERVICE_CONTROLLER_CONNECT) {
       return false;
     } else {
-      PANIC(
-          fmt::format("Could not dispatch Envoy to start as a service with error {}", last_error));
+      std::string error_msg{
+          fmt::format("Could not dispatch Envoy to start as a service with error {}", last_error)};
+      service_static->windows_event_logger_.log(::spdlog::source_loc{__FILE__, __LINE__, __func__},
+                                                spdlog::level::err, error_msg);
```
Comment (mattklein123) [inline source/exe/win32/service_base.cc:56]: It seems like we could define some macro here even if not the normal ones for this so we avoid the repetition?

File: source/exe/win32/service_base.h [hunk 2]
```diff
@@ -9,6 +9,8 @@
 
 #include "envoy/event/signal.h"
 
+#include "common/common/event_logger_impl.h"
```
Comment (mattklein123) [inline source/exe/win32/service_base.h:12]: nit: I'm not sure I would bother with the strip prefix on this include given this is really win32 specific and included from a win32 specific location?

File: source/common/common/win32/event_logger_impl.h [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+#define GENERATE_WIN32_NATIVE_LOGGER(X) WindowsEventLogger(#X),
+
+/**
+ * Logger that uses `spdlog::sinks::win_eventlog_sink`.
+ */
+class WindowsEventLogger : public Logger {
+public:
+  WindowsEventLogger(const std::string& name);
+
+  friend class ServiceBase;
```
Comment (mattklein123) [inline source/common/common/win32/event_logger_impl.h:17]: Why is this needed?

File: source/common/common/base_logger.h [hunk 4]
```diff
@@ -35,6 +36,12 @@ class Logger {
   void setLevel(spdlog::level::level_enum level) { logger_->set_level(level); }
   spdlog::level::level_enum level() const { return logger_->level(); }
 
+  template <typename... Args>
+  void log(spdlog::source_loc loc, spdlog::level::level_enum lvl, absl::string_view fmt,
+           const Args&... args) {
+    logger_->log(loc, lvl, fmt, args...);
+  }
```
Comment (mattklein123) [inline source/common/common/base_logger.h:43]: nit: comment please.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks LGTM with small comments.

/wait

**Comment by davinci26** [comment]
@mattklein123 sorry for the delayed response, I was on PTO. Ack the comments and I will address them today

File: source/exe/win32/service_base.cc (hunk 1 above)
Comment (davinci26) [reply source/exe/win32/service_base.cc]: General aversion towards C style macro's but I guess they are needed here to preserve the source location. added in b82744e1e064c95fa55788297120929d835372a8

File: source/common/common/win32/event_logger_impl.h (hunk 3 above)
Comment (davinci26) [reply source/common/common/win32/event_logger_impl.h]: I thought it was needed because of a TODO, but it turns out it is not. I removed it in It is not needed, I thought it was needed as I was experimenting

**Comment by davinci26** [comment]
Unrelated flake in coverage

**Review by mattklein123 — APPROVED** [review]
Thanks!
